### ISSUE TITLE
Websocket client works, server fails

### ISSUE THREAD

**Issue by tjacobs** [machinezone/IXWebSocket#267]
I'm using IXWebSocket (latest version, just cloned and built from source), and when I connect to a websocket server and sendBinary() as a client, it works fine.

But when I run a WebSocketServer, and sendBinary() the same way, it only sends some of it, and then the pipe breaks after a few seconds.

```
// Websocket server
int port = 8080;
string host = "192.168.1.1";
ix::WebSocketServer webSocketServer(port, host);

// Websocket client we send data back to
ix::WebSocket* webSocket = &webSocketClient;

___

        // -- Start Server --

        // Set messages callback
        webSocketServer.setOnClientMessageCallback(messageServer);

        // Listen
        auto response = webSocketServer.listen();
        if (!response.first) return false;

        // Start
        webSocketServer.start();

___

// Get message from WebSocket client
void message(const ix::WebSocketMessagePtr& msg);
void messageServer(std::shared_ptr<ix::ConnectionState> connectionState, ix::WebSocket& webSocketIn, const ix::WebSocketMessagePtr& msg)
{
        if (msg->type == ix::WebSocketMessageType::Open)
        {
                printf("New websocket client\n"); fflush(stdout);
                webSocket = &webSocketIn;
        }
        else if (msg->type == ix::WebSocketMessageType::Message)
        {
                printf("Message: %s\n", msg->str.c_str()); fflush(stdout);
                message(msg);
        }
}

....
sending...

                ix::WebSocketSendInfo sent = webSocket->sendBinary(header + sendBuffer);
                bytesSent += sent.wireSize;
                if (sent.wireSize < 0) cout << "Error sending: " << sent.wireSize << endl;
                else if (sent.wireSize != length+header.size()) cout << "Only sent: " << sent.wireSize-header.size() << " of " << length << endl;

result every time:
New websocket client
Message: web82368 LIST
....
Only sent: 13089 of 13121
Only sent: 14618 of 14589
Only sent: 16267 of 16284
Only sent: 17458 of 17372
Only sent: 12734 of 12668
Only sent: 10930 of 10874
Only sent: 60 of 2048
Only sent: 9448 of 9397
Only sent: 64 of 2048
Only sent: 88 of 2048
Only sent: 9444 of 9393
...
error:Broken pipe
```

This is running on a Pi.

Any ideas? Is it a performance thing? Sending to server with the client takes about 25% CPU, but I see it at 60% when it is a server and fails.

It's used in this by the way: http://teleportconnect.com/

**Comment by tjacobs** [comment]
Oh - it's the compression that is making it break - probably overloading the CPU, and it only happens when the IXWebSocket Pi is the server, not the client, because it can dictate the deflate options I guess.

I tried adding DeflateOptions deflate false, but it still is compressing the outgoing sendBinary() messages - any idea how to turn off the compression of outgoing messages?:

```
        if (msg->type == ix::WebSocketMessageType::Open)
        {
                printf("New websocket client\n"); fflush(stdout);
                webSocket = &webSocketIn;

                bool enablePerMessageDeflate = false;
                ix::WebSocketPerMessageDeflateOptions webSocketPerMessageDeflateOptions(enablePerMessageDeflate, false, false, 15, 15);
                webSocket->setPerMessageDeflateOptions(webSocketPerMessageDeflateOptions);
```

**Comment by tjacobs** [comment]
I tried adding:

```
        webSocketServer.disablePerMessageDeflate();
```

Yet the sent messages are still getting compressed (and often increasing in size because it's compressed video data already):

```
WireSizeA: 29
WireSizeC: 29 Chunk: 32768
WireSizeS: 29 Chunk: 32768
New websocket client
Message: web737673 LIST
WireSizeA: 17
Compressed: 19
WireSizeC: 19 Chunk: 32768
WireSizeS: 19 Chunk: 32768
WireSizeA: 21
Compressed: 11
WireSizeC: 11 Chunk: 32768
WireSizeS: 11 Chunk: 32768
WireSizeA: 24
Compressed: 12
WireSizeC: 12 Chunk: 32768
WireSizeS: 12 Chunk: 32768
...
--> Starting video.
WireSizeA: 21
Compressed: 14
WireSizeC: 14 Chunk: 32768
WireSizeS: 14 Chunk: 32768

Sending 13365 bytes of video
WireSizeA: 13365
Compressed: 13330
WireSizeC: 13330 Chunk: 32768
WireSizeS: 13330 Chunk: 32768
Only sent: 13322 of 13357

Sending 14680 bytes of video
WireSizeA: 14680
Compressed: 14753
WireSizeC: 14753 Chunk: 32768
WireSizeS: 14753 Chunk: 32768
Only sent: 14745 of 14672

Sending 16102 bytes of video
WireSizeA: 16102
Compressed: 16183
WireSizeC: 16183 Chunk: 32768
WireSizeS: 16183 Chunk: 32768
Only sent: 16175 of 16094
```

**Comment by tjacobs** [comment]
Any idea why deflate is still on after turning it off @bsergean ?

**Comment by tjacobs** [comment]
Ok I just made the change to disable deflate manually in the library and it works fine now. Closing.

```
WebSocketSendInfo WebSocketTransport::sendBinary(const std::string& message,
                                                     const OnProgressCallback& onProgressCallback)

    {
        return sendData(
            wsheader_type::BINARY_FRAME, message, false, /*_enablePerMessageDeflate,*/ onProgressCallback);
    }

    WebSocketSendInfo WebSocketTransport::sendText(const std::string& message,
                                                   const OnProgressCallback& onProgressCallback)

    {
        return sendData(
            wsheader_type::TEXT_FRAME, message, false, /*_enablePerMessageDeflate,*/ onProgressCallback);
    }
```

**Comment by bsergean** [comment]
Sorry my company has changed the way notifications work and I never receive them, so I discover issues 2 months too late. Cool project.

Happy you got something working. The client should dictate the compression, and call client.disablePerMessageDeflate(); / this should prevent the server from compressing, but it's a bug that the server exposes a disablePerMessageDeflate option which isn't working, I'll try to fix that when I get a chance.

**Comment by bsergean** [comment]
You can also build with -DUSE_ZLIB=OFF if you want to remove all compression code.

**Comment by bsergean** [comment]
The problem should be fixed in 0813eb1788e6660c0b0b7de1463c21d44eafc577

So you don't need to modify the library yourself anymore.
